os/fs/driver/mtd/smart.c

```diff
@@ -4308,8 +4306,11 @@ static int smart_writesector(FAR struct smart_struct_s *dev, unsigned long arg)
 
 		/* Since we performed a relocation, do garbage collection to
 		 * ensure we don't fill up our flash with released blocks.
+		 * Check if the number of free sectors is less than a certain threshold (4%)
 		 */
-		smart_garbagecollect(dev);
+		if (dev->freesectors < (dev->availSectPerBlk * dev->neraseblocks) / 25) {
```

Comment:
it's not good condition, we can adjust availSectPerBlk greater than 25.
I think making macro for condition will be better.
```
#define GC_CONDITION(dev) (dev->freesectors <= (dev->sectorsPerBlk + 4) || (dev->releasesectors > dev->freesectors && dev->freesectors < (dev->totalsectors >> 5)) ......
```

Reply:
The factor of 25 is not based on dev->availSectPerBlk
It represents 4% of the total sectors.

GC Threshold conditions are: -
1. (freesectors < (totalsectors / 32)) && (freesectors < releasedsectors)
This means freesectors < ~3.2% of totalsectors
2. freesectors < (dev->availSectPerBlk + 4)

A condition check of 4% is a superset of both of these conditions.
Hence, it limits the number of calls to the garbagecollect function without inhibiting any of the necessary calls.

Do you suggest that we should check both the threshold conditions itself with a MACRO before calling smart_garbagecollect()?

Reply:
![Screenshot_2022-10-20-16-58-46-516.png](https://user-images.githubusercontent.com/54265242/196936700-6db7d7c4-72a0-4b9e-89a4-f5b5e9adbf3a.png)

I think 5% is a better threshold to use, it covers our cases better

Reply:
@amogh-samsung What is the purpose of this modification? Is it to reduce the number of call to garbage collection? In the garbage collection fuction, there is already 4% check logic, so this modification can't change any behavior to garbage collection logic and wear leveling performance. Am I right?

Reply:
Yes, changing GC logic risks our feature and stability.
This proposition only limits the calls to GC function, does not change GC logic itself.

https://github.com/Samsung/TizenRT/pull/5633#issuecomment-1243535626